crates/matrix-sdk-crypto/src/machine/tests/megolm_sender_data.rs

```diff
@@ -0,0 +1,229 @@
+/*
+Copyright 2024 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{fmt::Debug, iter, pin::Pin};
+
+use assert_matches::assert_matches;
+use futures_core::Stream;
+use futures_util::{FutureExt, StreamExt};
+use matrix_sdk_test::async_test;
+use ruma::{room_id, user_id, RoomId, TransactionId, UserId};
+use serde::Serialize;
+use serde_json::json;
+
+use crate::{
+    machine::{
+        test_helpers::get_machine_pair_with_setup_sessions_test_helper,
+        tests::to_device_requests_to_content,
+    },
+    olm::{InboundGroupSession, SenderData},
+    store::RoomKeyInfo,
+    types::events::{room::encrypted::ToDeviceEncryptedEventContent, EventType, ToDeviceEvent},
+    EncryptionSettings, EncryptionSyncChanges, OlmMachine, Session,
+};
+
+/// Test the behaviour when a megolm session is received from an unknown device,
+/// and the device keys are not in the to-device message.
+#[async_test]
+async fn test_receive_megolm_session_from_unknown_device() {
+    // GIVEN Bob does not know about Alice's device
```

Comment:
I have been writing "Given" etc. without all-caps. If you prefer this way that's fine, but just in case you misremembered what I have been doing.